### PR TITLE
Return the correct status code when authenticating using the oauth se…

### LINKED ISSUE #9346: [API] Incorrect HTTP response code sent when username / password not provided for token retrieval
### Issue Summary

When POSTing `/ghost/api/v0.1/authentication/token` with an empty password, a 500 error is thrown. A 401 error would probably be better, since this is a client error rather than a server error, and a 401 is thrown when an empty client_id / client_secret is sent.

Form Data (as interpreted as ghost; I added a middleware function to log `req.body` to the console):

```
{ grant_type: 'password',
  username: 'bot@your-awesome.blog',
  password: '',
  client_id: 'ghost-test',
  client_secret: 'secret' }
```
Note: The empty password was provided because of a typo
[Another] Note: A similar error occured when username was not provided

<details>
<summary> Error Log</summary>

```
ERROR [2017-12-18 01:48:37] "POST /ghost/api/v0.1/authentication/token" 500 77ms

NAME: InternalServerError
CODE: invalid_request
MESSAGE: Missing required parameter: password

level:normal

InternalServerError: Missing required parameter: password
    at new GhostError ({ghost}/versions/1.19.0/core/server/lib/common/errors.js:9:26)
    at prepareError ({ghost}/versions/1.19.0/core/server/web/middleware/error-handler.js:39:19)
    at Layer.handle_error ({ghost}/versions/1.19.0/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix ({ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:315:13)
    at {ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:284:7
    at Function.process_params ({ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:335:12)
    at next ({ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:275:10)
    at Layer.handle_error ({ghost}/versions/1.19.0/node_modules/express/lib/router/layer.js:67:12)
    at trim_prefix ({ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:315:13)
    at {ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:284:7
    at Function.process_params ({ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:335:12)
    at next ({ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:275:10)
    at {ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:635:15
    at next ({ghost}/versions/1.19.0/node_modules/express/lib/router/index.js:260:14)
    at next ({ghost}/versions/1.19.0/node_modules/express/lib/router/route.js:127:14)
    at {ghost}/versions/1.19.0/node_modules/oauth2orize/lib/middleware/token.js:57:25

TokenError: Missing required parameter: password
    at Object.password [as handle] ({ghost}/versions/1.19.0/node_modules/oauth2orize/lib/exchange/password.js:90:32)
    at next ({ghost}/versions/1.19.0/node_modules/oauth2orize/lib/server.js:419:15)
    at Server._exchange ({ghost}/versions/1.19.0/node_modules/oauth2orize/lib/server.js:427:3)
    at token ({ghost}/versions/1.19.0/node_modules/oauth2orize/lib/middleware/token.js:56:12)
    at generateAccessToken ({ghost}/versions/1.19.0/core/server/auth/oauth.js:182:35)
    at Layer.handle [as handle_request] ({ghost}/versions/1.19.0/node_modules/express/lib/router/layer.js:95:5)
    at next ({ghost}/versions/1.19.0/node_modules/express/lib/router/route.js:137:13)
    at authenticate ({ghost}/versions/1.19.0/core/server/auth/authenticate.js:68:24)
    at Strategy.strategy.success ({ghost}/versions/1.19.0/node_modules/passport/lib/middleware/authenticate.js:214:18)
    at verified ({ghost}/versions/1.19.0/node_modules/passport-oauth2-client-password/lib/strategy.js:50:10)
    at then ({ghost}/versions/1.19.0/core/server/auth/auth-strategies.js:24:32)
    at tryCatcher ({ghost}/versions/1.19.0/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler ({ghost}/versions/1.19.0/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise ({ghost}/versions/1.19.0/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 ({ghost}/versions/1.19.0/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises ({ghost}/versions/1.19.0/node_modules/bluebird/js/release/promise.js:693:18)

```

</details>

### Technical details:

* Ghost Version: 1.19.0
* Node Version: 6.12.2
* Browser/OS: got (node request lib) / Windows subsystem for Linux
* Database: sqlite3

### PR DESCRIPTION
**Pull request by hugomarisco**
…rver

closes #9346
- attached the status code yielded by the oauth lib to the correct error attribute (.statusCode)

### CODE REVIEW

**Comment by hugomarisco** [comment]
This is my first contribution to Ghost. Let me know if anything needs to be changed/improved.

**Review by kirrg001 — CHANGES_REQUESTED** [review]
Hey! Thanks for your PR.

The suggested fix is not really the right one.

What happens is:
The oauthorize lib returns a none Ghost error.
And our Ghost middleware error handler transforms this error into a Ghost error (InternalServerError).
This ends in the situation:

```
name: InternalServerError
statusCode: 400
```

What i would suggest is to extend the error handler fn `_private.prepareError` to differentiate between 404 and 400/422. 404 handling exists already.

```

// prep
if (err.status) {
  err.statusCode = err.status;
}

else if (statusCode is 400 || statusCode is 422)
  new common.errors.ValidationError({err: err, message: err.message})
```

**Comment by hugomarisco** [comment]
Hello @kirrg001 , thanks for the code review 😃 

I understand the need of instantiating the error as a `ValidationError` but shouldn't that be done in the `oauth.js` file instead of `prepareError` function?

That way we keep the error handling middleware as agnostic as possible.

**Comment by kirrg001** [comment]
Fine with me 👍

**Comment by hugomarisco** [comment]
@kirrg001 I've updated the commit (squashed both commits for readability after the merge, not sure if that's the policy on this repo or not) to include the changes you mentioned.

I also wrote a couple of unit tests to cover this scenario.

Shout if anything still needs to be improved.

File: core/server/services/auth/oauth.js [hunk 1]
```diff
@@ -179,7 +179,13 @@ oauth = {
             accessToken: authUtils.getBearerAutorizationToken(req)
         };
 
-        return oauthServer.token()(req, res, next);
+        return oauthServer.token()(req, res, function (err) {
+            if (err && err.status === 400) {
+                err = new common.errors.ValidationError({err: err, message: err.message});
```
Comment (kirrg001) [inline core/server/services/auth/oauth.js]: It's probably better to return a BadRequestError - if a param is missing, it's syntactically wrong.

Comment (hugomarisco) [reply core/server/services/auth/oauth.js]: Updated

**Review by kirrg001 — APPROVED** [review]

**Pull request merged**